src/main/scala/com/yotpo/metorikku/test/Tester.scala

```diff
@@ -74,162 +74,407 @@ case class Tester(config: TesterConfig) {
-      if (metricExpectedResultRows.length == metricActualResultRows.length) {
-        for ((metricActualResultRow, rowIndex) <- metricActualResultRows.zipWithIndex) {
-          val mapOfActualRow = metricActualResultRow.getValuesMap(metricActualResultRow.schema.fieldNames)
-          val matchingExpectedMetric = matchExpectedRow(mapOfActualRow, metricExpectedResultRows)
-          if (Option(matchingExpectedMetric).isEmpty) {
-            errorsIndexArr = errorsIndexArr :+ rowIndex
-            errors = errors :+ s"[$metricName - $tableName] failed on row ${rowIndex + 1}: " +
-              s"Didn't find any row in test_settings.json that matches ${mapOfActualRow}"
+      val allExpectedFieldsTable = allExpectedFields.getOrElse(tableName, List[String]())
+      val actualResults = extractTableContents(job.sparkSession, tableName, config.test.outputMode.get)
+      val (expectedResults, actualResultsMap) = (metricExpectedTests(tableName), TestUtil.getMapFromDf(actualResults))
+      val (expectedResultsObjects, actualResultsObjects) = (TestUtil.getRowObjectListFromMapList(expectedResults),
+                                                            TestUtil.getRowObjectListFromMapList(actualResultsMap))
+      val colToMaxLengthValMap = TestUtil.getLongestValueLengthPerKey(expectedResults ++ actualResultsMap)
+      val tableConfiguredKeys = getConfiguredKeysByTableName(configuredKeys, tableName)
+      val tableKeys = tableConfiguredKeys match {
+        case Some(configuredKeys) => //defined
+          getInvalidConfiguredKeysTable(configuredKeys, allExpectedFieldsTable) match {
+            case Some(invalidKeys) => return getInvalidKeysNonExistingTableErrors(allExpectedFields, invalidKeys, tableName)
+            case _ => log.info(s"[$metricName - $tableName]: Configured key columns for ${tableName}: [${configuredKeys.mkString(", ")}]")
+              configuredKeys //valid keys
           }
-          else {
-            metricExpectedResultRows = metricExpectedResultRows.filter(_ != matchingExpectedMetric)
+        case _ => log.warn(s"[metricName - $tableName]: Hint: Define key columns for ${tableName} for better performance")
+          allExpectedFieldsTable //undefined keys
+      }
+      val (expectedKeys, actualKeys) = (KeyColumns.getKeyListFromMap(expectedResults, tableKeys), KeyColumns.getKeyListFromDF(actualResults, tableKeys))
+      val (expectedResultsDuplications, actualResultsDuplications) = (TestUtil.getKeyToIndexesMap(expectedKeys), TestUtil.getKeyToIndexesMap(actualKeys))
+      val (printableExpectedResults, printableActualResults) = (TestUtil.addLongestWhitespaceRow(expectedResultsObjects, colToMaxLengthValMap),
+                                                                 TestUtil.addLongestWhitespaceRow(actualResultsObjects, colToMaxLengthValMap))
+      val (whitespaceRowExpIndex, whitespaceRowActIndex) = (printableExpectedResults.length - 1, printableActualResults.length - 1)
```

Comment:
Seems weird... It's always the last line, just figure it out anytime you need it.